src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -344,15 +344,9 @@ export class MeetingManager implements AudioVideoObserver {
   async listAndSelectDevices(): Promise<void> {
     await this.updateDeviceLists();
 
-    let hasAudioInput = false, hasAudioOutput = false, hasVideoInput = false;
-    const devices = await navigator.mediaDevices.enumerateDevices();
-    for (const device of devices) {
-      if (device.label) {
-        if (device.kind === "audioinput") hasAudioInput = true;
-        if (device.kind === "audiooutput") hasAudioOutput = true;
-        if (device.kind === "videoinput") hasVideoInput = true;
-      }
-    }
+    const hasAudioInput = this.audioInputDevices?.some(value => value.label) || false;
+    const hasAudioOutput = this.audioOutputDevices?.some(value => value.label) || false;
+    const hasVideoInput = this.videoInputDevices?.some(value => value.label) || false;
```

Comment:
Does firefox sometimes not have the 'kind' field? 

How does this change the value of these three variables from the previous version? It looks cleaner and it reduces one call to enumerateDevices, but wouldn't the result be the same?

Reply:
This bug is quite tricky. 

* The `enumerateDevices()` in original  line 348 doesn't `await` the `getUserMedia()` call from `updateDeviceLists` and `DeviceProvider`
* So the `hasAudioInput`, `hasAudioOutput` and `hasVideoInput` keep a `false` status 
* After user navigates to `DevicePage`, although the devices are listed there, but none of them is selected. 
* This cause no video preview and no microphone activity in `DevicePage`. 
* After user joins the meeting, the selected device for audio input and video input are 'None'.

In this PR, I replace the
```ts
const devices = await navigator.mediaDevices.enumerateDevices();
```
with 
```ts
this.audioInputDevices, this.audioOutputDevices, this.videoInputDevices
```
These device info are update by `this.updateDevicesLists`, so the status of `hasAudioInput`, `hasAudioOutput` and `hasVideoInput` are consistency with the actual status.